proto-lens-tests/tests/names_test.hs

```diff
@@ -4,25 +4,115 @@
 -- license that can be found in the LICENSE file or at
 -- https://developers.google.com/open-source/licenses/bsd
 
--- This module tests that we can use various edge-case field names.
+-- This module checks our translation from proto names (fields and messages)
+-- into Haskell names.
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 module Main where
 
-import Data.ProtoLens (def)
-import Lens.Family2 ((&), (.~), (^.))
-import Prelude hiding (Maybe, maybe, head, span)
+import Data.Int (Int32)
+import Data.ProtoLens (def, Message)
+import Lens.Family2 (Lens', (&), (.~), (^.))
+import Prelude hiding (Maybe, maybe, map, head, span)
+import qualified Prelude
+import Test.Framework (Test, testGroup)
 import Test.Framework.Providers.HUnit (testCase)
 import Test.HUnit ((@=?))
 
 import Proto.Names
-import TestUtil (testMain)
-
-main = testMain [testNames]
-
-testNames = testCase "testPreludeNames" $ do
-    1 @=? (def & head .~ 1 :: Names) ^. head
-    1 @=? (def & span .~ 1 :: Names) ^. span
-    1 @=? (def & multiWordName .~ 1 :: Names) ^. multiWordName
-    1 @=? (def & fooMultiWORDName .~ 1 :: Names) ^. fooMultiWORDName
-    1 @=? (def & _multiWordName .~ 1 :: Names) ^. _multiWordName
-    1 @=? (def & type' .~ 1 :: Names) ^. type'
-    1 @=? (def & maybe .~ 1 :: Maybe) ^. maybe
+import TestUtil
+    ( TypedTest
+    , readFrom
+    , roundTripTest
+    , runTypedTest
+    , testMain
+    )
+
+main = testMain
+    [ testNames
+    , testPreludeType
+    , testHaskellKeywords
+    , testProtoKeywords
+    , testProtoKeywordTypes
+    , testReadReservedName
+    ]
+
+-- | Test that we can get/set each individual field.
+testFields :: forall a . (Show a, Message a, Eq a)
+           => String -> a
+           -> [SomeLens a Int32]
+           -> Test
+testFields name defValue fields = testGroup name
+    [ testCase "fields" $ mapM_ testField fields
+    , runTypedTest (roundTripTest "roundTrip" :: TypedTest a)
+    ]
+  where
+    testField (SomeLens f) = 1 @=? (defValue & f .~ 1) ^. f
+
+-- | Wraps a Lens' (which is a higher-order type) so it can be used in a list
+-- without ImpredicativeTypes.
+data SomeLens a b = SomeLens (Lens' a b)
```

Comment:
Done.  I was mixing up these higher-order types (where newtype works fine) for existential types (where it doesn't work).